sql/2020/03_Markup/pages_almanac_by_device.sql

```diff
@@ -0,0 +1,111 @@
+  contains_videos_without_autoplay BOOL,
+  html_node_lang STRING
+> LANGUAGE js AS '''
+var result = {};
+try {
+    var almanac = JSON.parse(almanac_string); 
+
+    if (Array.isArray(almanac) || typeof almanac != 'object') return result;
+
+    if (almanac.scripts) {
+      result.scripts_total = almanac.scripts.total;
+      if (almanac.scripts.nodes) {
+        result.none_jsonld_scripts_total = almanac.scripts.nodes.filter(n => !n.type || n.type.trim().toLowerCase() !== 'application/ld+json').length;
+        result.src_scripts_total = almanac.scripts.nodes.filter(n => n.src && n.src.trim().length > 0).length;
+
+        result.inline_scripts_total = result.none_jsonld_scripts_total - result.src_scripts_total;
+      }
+    }
+
+    if (almanac.headings_order) {
+      var good = true;
+      var previousLevel = 0;
+      almanac.headings_order.forEach(level =>  {
+          if (previousLevel + 1 < level) { // jumped a level
+            good = false;
+          }
+          previousLevel = level;
+      }); 
+      result.good_heading_sequence = good;
+    }
+
+    if (almanac.videos) {
+      var autoplay_count = almanac.videos.nodes.filter(n => n.autoplay == "" || n.autoplay).length; // valid values are blank or autoplay. Im just checking it exists...
+      //var autoplay_count = almanac.videos.nodes.filter(n => n.autoplay == "" || n.autoplay == "autoplay").length; 
+
+      result.contains_videos_with_autoplay = autoplay_count > 0;
+      result.contains_videos_without_autoplay = almanac.videos.total > autoplay_count;
+    }
+
+    if (almanac.html_node) {
+      result.html_node_lang = almanac.html_node.lang;
+    }
+
+} catch (e) {}
+return result;
+''';
+
+SELECT
+  client,
+  COUNT(0) AS total,
+
+  # has scripts excluding jsonld ones
+  AS_PERCENT(COUNTIF(almanac_info.none_jsonld_scripts_total > 0), COUNT(0)) AS pct_contains_none_jsonld_scripts_m204, 
+
+  # has inline scripts 
+  AS_PERCENT(COUNTIF(almanac_info.inline_scripts_total > 0), COUNT(0)) AS pct_contains_inline_scripts_m206,
+
+  # has src scripts 
+  AS_PERCENT(COUNTIF(almanac_info.src_scripts_total > 0), COUNT(0)) AS pct_contains_src_scripts_m208,
+
+  # has no scripts 
+  AS_PERCENT(COUNTIF(almanac_info.scripts_total = 0), COUNT(0)) AS pct_contains_no_scripts_m210,
+
+  # Does the heading logical sequence make any sense
+  AS_PERCENT(COUNTIF(almanac_info.good_heading_sequence), COUNT(0)) AS pct_good_heading_sequence_m222,
+
+  # pages with autoplaying video elements M310
+  AS_PERCENT(COUNTIF(almanac_info.contains_videos_with_autoplay), COUNT(0)) AS pct_contains_videos_with_autoplay_m310,
+
+  # pages without autoplaying video elements M311
+  AS_PERCENT(COUNTIF(almanac_info.contains_videos_without_autoplay), COUNT(0)) AS pct_contains_videos_without_autoplay_m311,
+
+  # pages with no html lang attribute M404
+  AS_PERCENT(COUNTIF(almanac_info.html_node_lang IS NULL OR LENGTH(almanac_info.html_node_lang) = 0), COUNT(0)) AS pct_no_html_lang_m404,
+
+FROM
+    ( 
```

Comment:
Nit: inconsistent indentation. Prefer 2 spaces for tabs like above.

Reply:
There's probably a few things like that. I keep tidying up when I spot things. A lot of the variation of formatting comes from where I stole the code from!

I personally let my development environment tidy up my own code. I've not worked out how to do that in Visual Studio Code for BigQuery yet. I also don't know all the formatting rules. Indentation is one thing that seemed inconsistent.